java/ql/src/Stubs/make_stubs.py

```diff
@@ -0,0 +1,155 @@
+# Tool to generate Java stubs for qltests
+
+import sys
+import os
+import subprocess
+import json
+
+
+def print_usage(exit_code=1):
+    print("Usage: python3 make_stubs.py testDir stubDir\n",
+          "testDir: the directory containing the qltest to be stubbed. Should contain an `options0` file pointing to the jars to stub, and an `options1` file pointing to `stubdir`\n",
+          "stubDir: the directory to output the generated stubs to")
+    exit(exit_code)
+
+
+if "--help" in sys.argv or "-h" in sys.argv:
+    print_usage(0)
+
+if len(sys.argv) != 3:
+    print_usage()
+
+testDir = sys.argv[1].rstrip("/")
+stubDir = sys.argv[2].rstrip("/")
+
+
+def check_dir_exists(path):
+    if not os.path.isdir(path):
+        print("Directory", path, "does not exist")
+        exit(1)
+
+
+def check_file_exists(path):
+    if not os.path.isfile(path):
+        print("File", path, "does not exist")
+        exit(1)
+
+
+def copy_file(src, dest):
+    with open(src) as srcf:
+        with open(dest, "w") as destf:
+            destf.write(srcf.read())
+
+
+check_dir_exists(testDir)
+check_dir_exists(stubDir)
+
+optionsFile = os.path.join(testDir, "options")
+options0File = os.path.join(testDir, "options0")
+options1File = os.path.join(testDir, "options1")
+
+check_file_exists(options0File)
+check_file_exists(options1File)
+
+# Does it contain a .ql file and a .java file?
+
+foundJava = False
+foundQL = False
+
+for file in os.listdir(testDir):
+    if file.endswith(".java"):
+        foundJava = True
+    if file.endswith(".ql") or file.endswith(".qlref"):
+        foundQL = True
+
+if not foundQL:
+    print("Test directory does not contain .ql files. Please specify a working qltest directory.")
+    exit(1)
+
+if not foundJava:
+    print("Test directory does not contain .java files. Please specify a working qltest directory.")
+    exit(1)
+
+
+javaQueries = os.path.abspath(os.path.dirname(sys.argv[0]))
+outputBqrsFile = os.path.join(testDir, 'output.bqrs')
+outputJsonFile = os.path.join(testDir, 'output.json')
+
+dbDir = os.path.join(testDir, os.path.basename(testDir) + ".testproj")
+
+
+def print_javac_output():
+    logDir = os.path.join(dbDir, "log")
+    if not os.path.isdir(logDir):
+        print("No database logs found")
+        return
+
+    logFile = None
+    for file in os.listdir(logDir):
+        if file.startswith("javac-output"):
+            logFile = os.path.join(logDir, file)
+            break
+    else:
+        print("No javac output found")
+
+    print("\nJavac output:\n")
+
+    with open(logFile) as f:
+        for line in f:
+            b1 = line.find(']')
+            b2 = line.find(']', b1+1)
+            print(line[b2+2:], end="")
+
+
+print("Stubbing qltest in", testDir)
+
+copy_file(options0File, optionsFile)
+
+cmd = ['codeql', 'test', 'run', '--keep-databases', testDir]
+print('Running ' + ' '.join(cmd))
```

Comment:
Try `shlex.join` to get quoting as necessary

Reply:
It looks like `subprocess.call` doesn't require quoting (and doesn't work when passed a whole string as the output of `shlex.join` rather than a list)

Reply:
oh yes sorry I should clarify, use shlex only for printing a command-line (this will render it neatly, and in a form that can be copy-pasted to another terminal for example)